libint: cmake build and SSE

sheepforce opened the pull request:
###### Motivation for this change
@markuskowa 

#132879 

Libint docs recommend to perform the code compilation step with cmake. This also installs the required CMake files for Psi4. 

There are some caveats currently:
  - the fortran interface does not build anymore. In principle `-DENABLE_FORTRAN` should do the trick, but here it complains about missing fields of an imported struct and does not build
  - it only works with a build within the source-tree (I also don't know why, but this alone isn't a catastrophe I would say). The actual problem is, that cmake and ninja don't want to do this by default and I've bent the `configurePhase` and `buildPhase` somehow to allow this. But it seems hacky.


###### Things done

Code compilation phase is executed by CMake now.
I've added an option to use AVX optimisations and enabled SSE2 by default, as all x86_64 CPUs should have at least SSE2.

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

symphorien commented on pkgs/development/libraries/libint/default.nix:
```diff
@@ -1,6 +1,6 @@
-, python3, perl, gmpxx, mpfr, boost, eigen, gfortran
-, enableFMA ? false
+, python3, perl, gmpxx, mpfr, boost, eigen, gfortran, cmake, ninja
+, enableFMA ? false, enableAVX ? false
```
```suggestion
, enableFMA ? false, enableAVX ? stdenv.hostPlatform.avxSupport
```
is maybe a better default ?

sheepforce replied:
Cool, didn't know that this exists. :partying_face: How exactly would this behave? Say if hydra builds this and the hydra worker has AVX enabled, would it go into the cache like this? This again would break this package for a lot of hardware.

symphorien replied:
No this flags will only be true if the user declaratively sets `localSystem` to a specific architecture/cpu model: https://nixos.wiki/wiki/Build_flags

sheepforce replied:
Thank you for pointing this out :slightly_smiling_face:

markuskowa commented on pkgs/development/libraries/libint/default.nix:
```diff
@@ -48,29 +50,54 @@ stdenv.mkDerivation rec {
     ./autogen.sh
+    ./configure --prefix=$(pwd) ${builtins.toString configureFlags}
+    make export -j $NIX_BUILD_CORES
+    tar -xf libint-${version}.tgz
```
At this point, I can see how splitting it up into two derivations could make sense:
1. Each of the derivations would be less hacky (?)
2. The AVX optimizations do not affect the code generation phase (?) and thus an optimized would be cheaper, when it only needs to recompile the generated code?
3. Would two derivations be more simple and readable?

sheepforce replied:
Yep, I think two are better. Maybe something like this:
```nix
let codeGen = stdenv.mkDerivation { .. };
    codeComp = stdenv.mkDerivation { .. };
in codeComp
```

I guess "yes" to all three points. But no idea what the Fortran problem is; I couldn't find anything similar in the libint issues and PRs. I will give it another try with 2 separate derivations.

sheepforce commented on pkgs/development/libraries/libint/default.nix:
```diff
@@ -48,29 +50,54 @@ stdenv.mkDerivation rec {
 
-  postInstall = ''
-    cp export/fortran/libint_f.mod $out/include/
+  buildPhase = ''
```
The nixpkgs CMake buildPhase does something like `mkdir build && cd build && cmake ../`. In this case this doesn't work for *some* reason (many includes suddenly missing). It works more or less fine with a in-source-tree build.

sheepforce commented on pkgs/development/libraries/libint/default.nix:
```diff
@@ -1,76 +1,114 @@
+    ];
+
+    # Can only build in the source-tree. A lot of preprocessing magic fails otherwise.
+    configurePhase = ''
```
Is there another (better) way to force cmake to do a build in the source tree?

markuskowa replied:
Set `dontUseCmakeBuildDir = true;` may work. (see https://github.com/NixOS/nixpkgs/blob/991eaaa024e5b52d20c676d30d4613a2553e01b7/pkgs/development/tools/build-managers/cmake/setup-hook.sh#L27)

sheepforce replied:
Thank you! That worked. :+1:

sheepforce commented:
**Note**: Definitely WiP: have changed the configure flags drastically to iterate faster.

Split into two derivations as discussed above. Still the Fortran compilations does not work and I can't figure out why. If I enter a nix-shell and build the tgz archive, that is generated by the first step, it works just fine. E.g.

```bash
bash > nix-shell -p stdenv cmake eigen boost python3 gfortran
nix-shell > tar -xvf libint-2.6.0.tgz && cd libint-2.6.0.tgz
nix-shell > cmake . -DENABLE_FORTRAN=ON
nix-shell > make -j 20
```
gives what one would expect. In the derivation again it does not: [libint.log](https://github.com/NixOS/nixpkgs/files/6961880/libint.log)

sheepforce commented on pkgs/development/libraries/libint/default.nix:
```diff
@@ -48,29 +50,54 @@ stdenv.mkDerivation rec {
-    cd export/fortran
-    make libint_f.o ENABLE_FORTRAN=yes
-    cd ../..
+  configurePhase = ''
```
Couldn't find a way to do it without. https://github.com/NixOS/nixpkgs/pull/133258#discussion_r685996977

SuperSandro2000 commented on pkgs/development/libraries/libint/default.nix:
```diff
@@ -1,76 +1,114 @@
+      runHook postConfigure
+    '';
+
+    # Parallel building messes up Fortran interface compilation
+    enableParallelBuilding = true;
```
Why is it then enabled?

sheepforce replied:
Still figuring out what is going on there with Fortran. I thought it was the parallel building, but actually it is not. Must be some other difference between manual compilation in a nix-shell and with `nix-build`. Leftover from my debugging purposes. Sorry for the confusion.

sheepforce commented:
Found the Fortran issue: it is the `LIBINT2_REALTYPE`. For some reason the Fortran interface and the vectorised instructions are incompatible. Either Fortran or vectorisation ... I will try to see if I can get the C part to build with vectorisation and the Fortran part without. Otherwise I will make this conditional. Not using vectorisation at all is probably a deal worse than introducing a conditional for Fortran.

sheepforce commented:
Working and ready for review now. A few notes:

- Fortran compilation is conditional, as we sacrifice SSE vectorisation for it
- Code generation and code compilation are separated now
- Compiling of at least the second derivation works in parallel
- Added missing eigen (was never in build inputs)
- CMake should be installed now and detectable by other CMake derivations

markuskowa commented on pkgs/development/libraries/libint/default.nix:
```diff
@@ -1,76 +1,103 @@
+    # Can only build in the source-tree. A lot of preprocessing magic fails otherwise.
+    dontUseCmakeBuildDir = true;
+
+    enableParallelBuilding = true;
```
I think this is the default already for `cmake`
```suggestion
```

markuskowa commented on pkgs/development/libraries/libint/default.nix:
```diff
@@ -1,76 +1,101 @@
   };
-}
+
+  codeGen = stdenv.mkDerivation rec {
```
```suggestion
  codeGen = stdenv.mkDerivation {
```

markuskowa commented on pkgs/development/libraries/libint/default.nix:
```diff
@@ -1,76 +1,101 @@
+    inherit meta;
+  };
+
+  codeComp = stdenv.mkDerivation rec {
```
```suggestion
  codeComp = stdenv.mkDerivation {
```

sheepforce committed:
libint: code generation and compilation separated, using cmake, fortran


fortran compilation by cmake


libint: more optimisations for SSE


libint: avx switching based ony system configuration


libint: split into 2 derivations


libint: conditional fortran


libint: parallel building by default


libint: remove rec statement from mkDerivation

markuskowa reviewed (approved):
This looks very good now. The cmake part now actually works with parallel building, which reduces the build time dramatically.
